internal/kafka/internal/services/quota/ams_quota_service.go

```diff
@@ -396,3 +398,79 @@ func (q amsQuotaService) GetSubscriptionByID(subscriptionID string) (*amsv1.Subs
 
 	return resp.Body(), true, nil
 }
+
+// Checks if an organisation has a SKU entitlement and that entitlement is still active in AMS
+func (q amsQuotaService) IsQuotaEntitlementActive(kafka *dbapi.KafkaRequest) (bool, error) {
+	kafkaBillingModel, err := q.kafkaConfig.GetBillingModelByID(kafka.InstanceType, kafka.ActualKafkaBillingModel)
+	if err != nil {
+		return false, err
+	}
+
+	amsRelatedResourceBillingModel, err := q.getAMSRelatedResourceBillingModel(kafka.SubscriptionId, kafkaBillingModel)
+	if err != nil || arrays.StringEmptyPredicate(amsRelatedResourceBillingModel) {
+		return false, err
+	}
+
+	orgID, err := q.amsClient.GetOrganisationIdFromExternalId(kafka.OrganisationId)
+	if err != nil {
+		return false, err
+	}
+
+	quotaCostRelatedResourceFilter := ocm.QuotaCostRelatedResourceFilter{
+		ResourceName: &kafkaBillingModel.AMSResource,
+		Product:      &kafkaBillingModel.AMSProduct,
+		BillingModel: &amsRelatedResourceBillingModel,
+	}
+	quotaCosts, err := q.amsClient.GetQuotaCosts(orgID, true, false, quotaCostRelatedResourceFilter)
+	if err != nil {
+		return false, err
+	}
+
+	// SKU not entitled to the organisation
+	if len(quotaCosts) == 0 {
+		glog.V(10).Infof("ams quota cost not found for Kafka instance %q in organisation %q: %+v", kafka.ID, orgID, quotaCostRelatedResourceFilter)
+		return false, nil
+	}
+
+	if len(quotaCosts) > 1 {
+		glog.Warning("more than 1 quota cost was returned with the following related resource filter %+v for organisation %q", orgID, quotaCostRelatedResourceFilter)
+	}
+
+	// result will always return one quota for the given org, ams related resource billing model, resource and product for rhosak quotas
+	quotaCost := quotaCosts[0]
+
+	// when a SKU entitlement expires in AMS, the allowed value for that quota cost is set back to 0
+	// if the allowed value is 0 and consumed is greater than this, that denotes that the SKU entitlement
+	// has expired and is no longer active.
+	if quotaCost.Allowed() == 0 && quotaCost.Consumed() > quotaCost.Allowed() {
+		glog.V(10).Infof("quota %q has expired for organisation %q", quotaCost.QuotaID(), orgID)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// Determine the AMS QuotaCost related resource billing model from the AMS Subscription reserved resource billing model
+// based on the given subscription id and kafka billing model.
+func (q amsQuotaService) getAMSRelatedResourceBillingModel(subscriptionID string, kafkaBillingModel config.KafkaBillingModel) (string, error) {
+	reservedResources, err := q.amsClient.GetReservedResourcesBySubscriptionID(subscriptionID)
+	if err != nil {
+		return "", err
+	}
+
+	for _, rr := range reservedResources {
+		if kafkaBillingModel.AMSResource == rr.ResourceName() && kafkaBillingModel.HasSupportForAMSBillingModel(string(rr.BillingModel())) {
```

Comment:
The check of `kafkaBillingModel.HasSupportForAMSBillingModel(string(rr.BillingModel()` doesn't seem correct.
HasSupportForAMSBillingModel checks for the billing models in the config which are AMS Reserved Resource Billing Models, and not AMS Related Resource billing models